File: c#/src/WOT.TAK.Connection/src/TCPConnector.cs
using System;
using System.IO;
using System.Net.Sockets;

namespace WOT.TAK.Connection
{
    public class TCPConnector : TAKServerConnector
    {
        private String _url;
        private int _port;
        private String _responseStoragePath;
        private TcpClient _socket;
        protected StreamReader _input;
        protected StreamWriter _output;
        private Stream _stream;
        private Thread _listener;
        private bool _run;
        public TCPConnector(String url, String port, String responseStoragePath)
        {
            _url = url;
            _port = Int32.Parse(port);
            _responseStoragePath = responseStoragePath;
            _socket = new TcpClient();
            _run = true;
        }

        public void SendFile(string fileName)
        {
            try
            {
                using (StreamReader sr = new StreamReader(fileName))
                {
                    string line;
                    while ((line = sr.ReadLine()) != null)
                    {
                        _output.WriteLine(line);
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }

        public void SetStream(Stream stream)
        {
            _stream = stream;
        }

        public void Connect()
        {
            try
            {
                _socket.Connect(_url, _port);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
            if (_stream == null)
            {
                _stream = _socket.GetStream();
            }
            _output = new StreamWriter(_stream);
            _input = new StreamReader(_stream);
            _socket.ReceiveTimeout = 1000;
            _listener = new Thread(new ThreadStart(ResponseListener));
            _listener.Start();
        }
        private void ResponseListener()
        {
            string line;
            while(_run)
            {
                try
                {
                    line = _input.ReadLine();
                    using (StreamWriter sw = File.CreateText(_responseStoragePath + "/" + DateTimeOffset.Now.ToUnixTimeMilliseconds() + ".cot"))
                    {
                        sw.WriteLine(line);
                    }
                }
                catch (SocketException e)
                {
                    continue;
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine(ex.Message);
                }
            }
        }
        public void Close()
        {
            _run = false;
            Thread.Sleep(200);
            _output.Close();
            _input.Close();
            _socket.Close();
        }

        public TcpClient GetSocket() { return _socket; }
    }
}

File: connection/src/main/java/org/wot/tak/connection/TCPConnector.java
package org.wot.tak.connection;

import javax.sql.rowset.spi.XmlReader;
import java.io.*;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.nio.file.Files;

public abstract class TCPConnector implements TAKServerConnector {
    public final String url;
    public final int port;
    private final String responseStoragePath;
    protected SocketFactory socketFactory;
    protected Socket s;
    protected InputStream in;
    protected PrintWriter out;
    private ResponseListener listener;

    protected TCPConnector(String url, String port, String responseStoragePath, SocketFactory sFactory) {
        this.url = url;
        this.port = Integer.parseInt(port);
        this.responseStoragePath = responseStoragePath;
        this.socketFactory = sFactory;
    }

    public void sendFile(File file) throws IOException{
        try (BufferedReader br = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = br.readLine()) != null) {
                out.println(line);
            }
        }
    }

    public void connect() throws IOException {
        this.initializeInOut();
        s.setSoTimeout(100);
        listener = new ResponseListener();
        listener.start();
    }

    private void initializeInOut() throws IOException {
        out = new PrintWriter(s.getOutputStream(), true);
        in = s.getInputStream();
    }

    private class ResponseListener extends Thread {

        private volatile boolean stop = false;

        protected void finish(){
            stop = true;
        }
        public void run(){
            String xml;
            File file;
            PrintWriter pw;
            while (!stop) {
                try {
                    xml = EventReader.readFrom(in);
                    file = new File(responseStoragePath + "/" + System.currentTimeMillis() + ".cot");
                    Files.createDirectories(file.getParentFile().toPath());
                    pw = new PrintWriter(new FileWriter(file));
                    pw.println(xml);
                    pw.close();
                } catch (SocketTimeoutException ex) {
                    continue;
                } catch (Exception ex) {
                    System.err.println(ex.getMessage());
                }
            }
        }
    }

    public void close() throws Exception {
        listener.finish();
        Thread.sleep(200);
        out.close();
        in.close();
        s.close();
    }
}
